qualtran/bloqs/arithmetic/comparison.py

```diff
@@ -1154,3 +1156,423 @@ def _clineardepthgreaterthan_example() -> CLinearDepthGreaterThan:
+        c = bb.join(c_arr)
+
+        c = bb.add(BitwiseNot(QUInt(dtype.bitsize)), x=c)
+        a, b = bb.add(
+            OutOfPlaceAdder(self.dtype.bitsize + 1, include_most_significant_bit=False).adjoint(),
+            a=a,
+            b=b,
+            c=c,
+        )
+        a = bb.add(Cast(dtype, QUInt(dtype.bitsize)).adjoint(), reg=a)
+        b = bb.add(Cast(dtype, QUInt(dtype.bitsize)).adjoint(), reg=b)
+        b = bb.add(BitwiseNot(dtype), x=b)
+
+        if isinstance(self.dtype, QInt):
+            a = bb.add(SignExtend(self.dtype, QInt(self.dtype.bitsize + 1)).adjoint(), x=a)
+            b = bb.add(SignExtend(self.dtype, QInt(self.dtype.bitsize + 1)).adjoint(), x=b)
+        else:
+            a_arr = bb.split(a)
+            a = bb.join(a_arr[1:])
+            b_arr = bb.split(b)
+            b = bb.join(b_arr[1:])
+            bb.free(a_arr[0])
+            bb.free(b_arr[0])
+        bb.free(target)
+        return {'a': a, 'b': b}
+
+    def build_composite_bloq(
+        self,
+        bb: 'BloqBuilder',
+        a: 'Soquet',
+        b: 'Soquet',
+        c: Optional['Soquet'] = None,
+        target: Optional['Soquet'] = None,
+    ) -> Dict[str, 'SoquetT']:
+        if self.uncompute:
+            # Uncompute
+            assert c is not None
+            assert target is not None
+            return self._uncompute(bb, a, b, c, target)
+        else:
+            assert c is None
+            assert target is None
+            return self._compute(bb, a, b)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> 'BloqCountDictT':
+        dtype = attrs.evolve(self.dtype, bitsize=self.dtype.bitsize + 1)
+        counts: 'BloqCountDictT' = {}
+        if isinstance(self.dtype, QUInt):
+            counts = {BitwiseNot(dtype): 3}
+        else:
+            counts = {BitwiseNot(dtype): 2, BitwiseNot(QUInt(dtype.bitsize)): 1}
+
+        counts[CNOT()] = 1
+
+        adder = OutOfPlaceAdder(self.dtype.bitsize + 1, include_most_significant_bit=False)
+        if self.uncompute:
+            adder = adder.adjoint()
+        counts[adder] = 1
+
+        return counts
+
+
+@frozen
+class _HalfComparisonBase(Bloq):
+    dtype: _SUPPORTED_DTYPES
+    comparison_dtype: str = attrs.field(
+        default='>', validator=lambda _, __, s: s in ('>', '<', '>=', '<=')
+    )
+    uncompute: bool = False
+
+    @cached_property
+    def signature(self) -> Signature:
+        side = Side.LEFT if self.uncompute else Side.RIGHT
+        return Signature(
+            [
+                Register('a', self.dtype),
+                Register('b', self.dtype),
+                Register('c', QUInt(bitsize=self.dtype.bitsize + 1), side=side),
+                Register('target', QBit(), side=side),
+            ]
+        )
+
+    def adjoint(self) -> '_HalfComparisonBase':
+        return attrs.evolve(self, uncompute=self.uncompute ^ True)
+
+    @cached_property
+    @abc.abstractmethod
+    def _half_greater_than_bloq(self) -> Bloq:
```

Comment:
why can't this just be an attribute of `_HalfComparison[Base]`

Reply:
it will differ between different falvours ... the ones I'm implementing here have linear depth and linear complexity, this is why their names start with LinearDepth.... There is another variant with log depth but nlog complexity which we might want to add later (we already have one instance of it in `LessThanEqual`)